### ISSUE TITLE
[HOLD PR #42892][$250] Text pasted from Slack won't send or disappears in NewDot.

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#39971]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.61-5
**Reproducible in staging?:** y
**Reproducible in production?:** n
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @mallenexpensify 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1712695994757859

## Action Performed:

Scenario 1:

1. Copy text from from Slack
2. Paste text in compose box
3. Tap return

Scenario 2:

1. Copy text from from Slack
2. Type > in compose box
3. Paste text after > (and a space after)
4. Tap return

## Expected Result:
Message is sent

## Actual Result:
Scenario 1: Message stays in compose box and won't send. Cursor stays at the beginning of the message
Scenario 2: Only the > part of the message sends.

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/38435837/e5d2ba57-5336-40ce-9aa7-6800505cc0e8

https://github.com/Expensify/App/assets/38435837/384d0321-b113-4ea0-8537-3fb1990be135

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01892e3b8e5931a268</li>
        <li>Upwork Job ID: 1778891800225918976</li>
        <li>Last Price Increase: 2024-04-12</li>
<li>Automatic offers: </li>
<ul>
<li>jayeshmangwani | Reviewer | 0</li>
<li>kpadmanabhan | Contributor | 0</li>
</ul></li>
    </ul>
</details>

**Comment by mountiny** [comment]
@roryabraham been discussing [here](https://expensify.slack.com/archives/C06BDSWLDPB/p1712697245830559) with @thienlnam that this is specific for Slack it seems so we should not have to hold deploy for this bug

**Comment by brkdavis** [comment]
@m-natarajan can you pls paste the unformatted content of the slack message? initial suspicion is `>` symbol is treated with a special character when you copy to clipboard.

**Comment by brkdavis** [comment]
Your Expensify account email: brkdavis890@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01cef163ba3c7ed327?mp_source=share

**Comment by mallenexpensify** [comment]
Also happens when trying to copy/paste from https://github.com/Expensify/App/blob/main/contributingGuides/HOW_TO_BECOME_A_CONTRIBUTOR_PLUS.md

**Comment by roryabraham** [comment]
making this external

**Comment by brkdavis** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Text pasted from many editors including TextEdit or Notepad with a triple click to the NewDot report composer is not enabling "Send" button. 

### What is the root cause of that problem?
When you triple click and copy a text to clipboard, it brings along hidden `CR` `LF` characters along with it. After pasting cursor goes to position zero.
NOT WORKING STRING
![image](https://github.com/Expensify/App/assets/166563016/fd2faefa-945f-41a2-95b9-3675ed5ccd48)

However, if you copy a line from start to end by drag and select and copy, the `CR` `LF` characters do not come along with it. After pasting, cursor goes to end of string.
WORKING STRING
![image](https://github.com/Expensify/App/assets/166563016/1241ca2d-4526-40b5-8539-e050476ed221)
In this case the Send button is enabled.
![image](https://github.com/Expensify/App/assets/166563016/29a15878-fdfe-49ac-ad8f-536388e3564f)

Drag and drop of multi-line text works fine as long as you copy till end of line and NOT include newline.
![image](https://github.com/Expensify/App/assets/166563016/363b4ef1-6414-4cfa-a08e-34bf0610feef)
WORKING STRING
![image](https://github.com/Expensify/App/assets/166563016/b8048356-796a-455c-b79e-4e623f11d53d)

Multi-line that include trailing `CR` `LF` is not working as well. Note that in this case cursor is in position 0.
![image](https://github.com/Expensify/App/assets/166563016/268ff73a-61d5-4542-916e-e1443fa33adf)

NOT WORKING STRING
![image](https://github.com/Expensify/App/assets/166563016/f57ba79e-f527-424c-b4c3-91d528d61265)


NOT WORKING STRING results in following to be TRUE in `ReportActionComposer.tsx`.
```
  var isSendDisabled = isCommentEmpty || isBlockedFromConcierge || !!disabled || hasExceededMaxCommentLength;
```
This is because `isCommentEmpty` returns TRUE. 
ONYX update in `ComposerWithSuggestions.tsx`  for NOT WORKING STRING is not fired through `onChangeText` event and hence draft report is not saved to ONYX. This is because `useHtmlPaste` hook handles paste action, but captures cursor position with startIndex as 0 and endIndex as 0 due to additional `CR` `LF` at the end of string. 


### What changes do you think we should make in order to solve the problem?
There are 2 ways to handle this issue.
1. Before pasting, remove the trailing carriage return and line feed in `useHtmlPaste` --> `index.ts` and paste the text without these, so that ONYX update can be fired as expected.
```
var handlePastePlainText = useCallback(function (event) {
    . . . 
    plainText = plainText.replace(/^\s+|\s+$/g, '');
    if (plainText) {
      paste(plainText);
    }
  }, [paste]);
```

2. Fix the index issue. Make sure that start and end index are calculated correctly in `useHtmlPaste` --> `index.ts`  by `insertAtCaret()` so that the paste is handled correctly and subsequent events are fired accordingly.
```
. . . 
    // Move caret to the end of the newly inserted text node.
    range.setStart(node, node.length);
    range.setEnd(node, node.length);
    // If the node picked up is only CR and LF, then move to end of previous node and then set the range.
. . . 
```

Both will involve a little extensive regression testing effort to make sure that no other components using the same hooks are affected by this change. 

### What alternative solutions did you explore? (Optional)
After pasting a NON WORKING STRING, move to end of string and programmatically add a white space at end of string. This will enable the "Send" button.

**Comment by jayeshmangwani** [comment]
Reviewing Proposal

**Comment by roitman-g** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Every time the user pastes anything with the new line in the end the cursor is incorrectly put in the beginning of the text, so if you were to triple click something copy and paste as well as copying and pasting anything with a new line this issue appears.

### What is the root cause of that problem?
The root problem is that in "insertAtCaret" function in useHTMLPaste hook the selection range start and end index is determined incorrectly in cases there is a new line in the end of the insertion. 

### What changes do you think we should make in order to solve the problem?
I suggest using document.execCommand('insertHTML', false, text); instead of implementing it by ourselves (insertAtCaret). Method execCommand is being deprecated, however it is already being used in insertByCommand function as well as "copy" functionality and it does not seem like there is a [valid alternative right now](https://stackoverflow.com/questions/60581285/execcommand-is-now-obsolete-whats-the-alternative). 

### What alternative solutions did you explore? (Optional)
N/A

**Comment by jayeshmangwani** [comment]
@roitman-g We can not replace `insertAtCaret` with `document.execCommand` altogether, as it was added intentionally to handle the other custom functionalities in the app; instead, can we fix in the`insertAtCaret` function that is causing this issue ?

**Comment by roitman-g** [comment]
@jayeshmangwani do you mean that dispatching a 'paste' and 'input' event is necessary? If that is the case

we can replace only these lines in insertAtCaret with execCommand. 
```
        const range = selection.getRangeAt(0);
        range.deleteContents();
        const node = document.createTextNode(text);
        range.insertNode(node);

        // Move caret to the end of the newly inserted text node.
        range.setStart(node, node.length);
        range.setEnd(node, node.length);
        selection.removeAllRanges();
        selection.addRange(range);
```

If not, could you say which kind of functionalities so I can test? I tried pasting all kind of things and it seems like besides the end of the line edge case, this piece of code works the same as execCommand('insertHTML, .....).

Also, there is [issue ](https://github.com/Expensify/App/issues/40033) which probably a duplicate, they have the same root cause.

**Comment by roitman-g** [comment]
@jayeshmangwani this case works the same for me with execCommand('insertHTML',  false, text). The result I get when pasting it:


<img width="538" alt="Снимок экрана 2024-04-15 в 23 29 27" src="https://github.com/Expensify/App/assets/23198681/0b64acae-68b4-4c23-80d7-cf6f8fdd130d">

<img width="582" alt="Снимок экрана 2024-04-15 в 23 30 37" src="https://github.com/Expensify/App/assets/23198681/5aaffa17-b189-405b-9a09-edaab98eafb8">

**Comment by brkdavis** [comment]
##Proposal 
[Updated](https://github.com/Expensify/App/issues/39971#issuecomment-2053860349)

**Comment by jayeshmangwani** [comment]
@brkdavis's [Updated proposal](https://github.com/Expensify/App/issues/39971#issuecomment-2053860349) looks good to me 

@brkdavis Just one suggestion(question): rather than removing all the whitespace from the beginning and end of a string, can we remove the carriage return and line feed characters only from the start and end like this `/^[\r\n]+|[\r\n]+$/g` ?

🎀 👀 🎀 C+ reviewed

**Comment by roitman-g** [comment]
there is also a bug worth mentioning though probably not related to the paste functionality - if you copy whatever is in the "copy the following text" manually (so not with copy icon button and it is copied as a code block) and paste it - you cannot send the message and chat box is not expanded.

> @roitman-g The Purpose of this logic was this PR #39177. Now, if you want to verify that your suggested changes will not recreate the issue mentioned in the PR, then you can follow the Tests section of PR after applying your suggested changes; I am copying the Tests from the PR below.
> 
> 1. Copy following text:
> 
> ```
> https://expensify.com
> # Lorem ipsum
> > Hello world
> `foo`
> @here
> 
> @someone@swmansion.com
> ```
> 
> 2. Enter any chat
> 3. Paste copied text from step 1 into main composer
> 4. Verify if cursors is positioned properly, it should be at the end of the pasted text
> 5. Verify if there are no extra new lines. Especially verify if the empty line is inside the text from step 1, hasn't been duplicated

**Comment by slafortune** [comment]
@roryabraham I will be out until 4/22, looks like the ball is in your court at the moment.  Not adding another BZ since this looks like it's in a good place.

**Comment by brkdavis** [comment]
> @brkdavis's [Updated proposal](https://github.com/Expensify/App/issues/39971#issuecomment-2053860349) looks good to me
> 
> @brkdavis Just one suggestion(question): rather than removing all the whitespace from the beginning and end of a string, can we remove the carriage return and line feed characters only from the start and end like this `/^[\r\n]+|[\r\n]+$/g` ?
> 
> 🎀 👀 🎀 C+ reviewed

Sure. I will test with removal of trailing CR LF alone and proceed if looking good.
@jayeshmangwani / @roryabraham : shall I proceed with a PR? but I still do not see an offer yet on upwork?

**Comment by jayeshmangwani** [comment]
@brkdavis do not proceed with PR, we still need approval from @roryabraham , and then you will get an upwork offer then you can start with the PR

**Comment by jayeshmangwani** [comment]
@brkdavis As you are a new contributor and its your first assigned issue according to our contributing guidelines `New contributors are limited to working on one job at a time`, please do not post proposal on any other issue until this issue got completed(PR hit production) and if you have already submitted proposal on any other issue and if your proposal selected then please let the c+ know that your first PR is in review.

**Comment by brkdavis** [comment]
Contributor details
Your Expensify account email: brkdavis890@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01063aad97ce0e22c6

@roryabraham : can you please help me link to my correct upwork account?

**Comment by jayeshmangwani** [comment]
@brkdavis You have linked your account with the below details; what's the issue?

Your Expensify account email: [brkdavis890@gmail.com](mailto:brkdavis890@gmail.com)
Upwork Profile Link: https://www.upwork.com/freelancers/~01cef163ba3c7ed327?mp_source=share

**Comment by brkdavis** [comment]
@jayeshmangwani : melvin-bot was not picking my upwork profile. I think it is linked to another email of mine.
I created a dummy upwork profile to confirm this. Unfortunately, now my account is linked to this dummy upwork profile.
I would like to use this upwork profile: https://www.upwork.com/freelancers/~01cef163ba3c7ed327?mp_source=share
How can this be corrected now?

**Comment by jayeshmangwani** [comment]
@brkdavis I don't know how to update the Upwork profile in melvin-bot; you can raise the issue on Slack in #expensify-open-source

**Comment by brkdavis** [comment]
@jayeshmangwani : I will do that. Or is re-assigning this issue to another account of mine faster?
https://github.com/kpadmanabhan

**Comment by kpadmanabhan** [comment]
> @jayeshmangwani : I will do that. Or is re-assigning this issue to another account of mine faster? https://github.com/kpadmanabhan

Can you please re-assign this to this account of mine?

**Comment by kpadmanabhan** [comment]
Contributor details
Your Expensify account email: krispv@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01063aad97ce0e22c6

**Comment by jayeshmangwani** [comment]
@roryabraham We have been getting some strange issues here, @brkdavis is having some issues with storing their Upwork profile data in melvin-bot. They have [another Github](https://github.com/Expensify/App/issues/39971#issuecomment-2064882143) and an [Upwork account](https://github.com/Expensify/App/issues/39971#issuecomment-2064889702); that account is linked correctly. Should we assign it to their new GH account @kpadmanabhan so we can move forward with PR?

**Comment by jayeshmangwani** [comment]
@kpadmanabhan What's your ETA for PR?

**Comment by kpadmanabhan** [comment]
@jayeshmangwani Code changes are good. Working on the unit test strategy for the change. 

@jayeshmangwani / @roryabraham : Does writing a test for the `string.replace(...)` add value? I dont think it does. 
Hence working on following strategies for unit tests / integration tests
- for `useHtmlPaste` hook. But this cannot cover if the send button will be enabled or not. This has to be checked in manual test or e2e automated test. Did you mean e2e test when you said unit tests? Facing challenges with heavy consumption of UI components (window and DOM) inside the hooks. 
- for `composer`. This will cover paste and send button by mocking other components around it. 

Or do you mean test the newly written regular expression alone? Either by mocking UI components or by extracting string sanitization into a different function and test that alone?

Please let me know your thoughts.

**Comment by roryabraham** [comment]
@kpadmanabhan thanks for putting some thought into an automated testing strategy. I think that tests in `useHtmlPaste` covering the removal or carriage returns might be sufficient, and including tests for `composer` that covers paste and send button would be above-and-beyond (would love to see it). Tests for `useHtmlPaste` can probably be classified as unit tests, while tests for composer might be more accurately classified as "ui tests", probably relying on react-native-testing-library for mocked renders.

We do not currently have any functional e2e tests, so we shouldn't try to establish a new pattern by adding them here.

**Comment by kpadmanabhan** [comment]
Thanks @roryabraham. 
Concept for ui tests for composer
```
        const pasted = jest.fn(() => null);
        const changed = jest.fn(() => null);
        
        render(<AnimatedMarkdownTextInputRef data-testid="composer-input" />);
        
        const composerElement = screen.queryByTestId('composer-input');
        useHtmlPaste(composerElement, handlePaste, true);

        const pasteFromClipboard = createEvent.paste(composerElement, {
            clipboardData: {
              getData: () => '123456',
            },
          });
        
          fireEvent(composerElement, pasteFromClipboard);
        
          expect(pasted).toHaveBeenCalled();
          expect(pasted).toHaveBeenCalledWith('123456');
```

We use `testing-library` for react, but this will need some changes to the component as well to support testability.

**Comment by roryabraham** [comment]
> We use testing-library/react

Let's use testing-library/react-native ?

**Comment by slafortune** [comment]
@kpadmanabhan do you have an update on this?

**Comment by kpadmanabhan** [comment]
@slafortune: There is no unit test that exists for the hooks in our codebase. `useHtmlPaste` uses `window.getSelection` as well as `Selection.getRange()`. Landing up into issues mocking these. I will use #expensify-open-source to get some help on this. Targetting PR on priority today.

**Comment by jayeshmangwani** [comment]
@kpadmanabhan Please raise a PR(or Draft) with the regex changes; we can discuss the tests on the PR.

**Comment by slafortune** [comment]
@kpadmanabhan @jayeshmangwani do you have an update?  Looks like the draft is waiting on you @kpadmanabhan .

**Comment by jayeshmangwani** [comment]
We await the update for the @kpadmanabhan; yesterday bumped on PR https://github.com/Expensify/App/pull/41265#issuecomment-2102701521

**Comment by roryabraham** [comment]
PR was abandoned, so we're looking to hire another contributor to take this one on.

**Comment by kpadmanabhan** [comment]
Sorry for that. 
Update:
https://github.com/Expensify/App/pull/41265#issuecomment-2118674202

**Comment by mallenexpensify** [comment]
Hope you're feeling better @kpadmanabhan .  Please post daily updates on the GH or PR til it's finished. Thx

**Comment by slafortune** [comment]
@jayeshmangwani  - C+ role ~ will be paid via New Dot $250
@kpadmanabhan  - Contributor role ~ will be due $250

I will be on vacation until June 4th so adding another BZ in the meantime.  I will take this back on the 4th.

**Comment by roryabraham** [comment]
@kpadmanabhan appears to have picked the PR back up

**Comment by kpadmanabhan** [comment]
> @kpadmanabhan appears to have picked the PR back up

yes @roryabraham . tested in all platforms and all artifacts are attached in PR template.
can you pls review?

**Comment by bfitzexpensify** [comment]
I'm heading ooo but Steve is due back on Tuesday, and I don't think any action will be required before then from the BZ team.

**Comment by BartoszGrajdek** [comment]
Hi, I’m Bartosz from SWM `react-native-live-markdown` team. We’ve recently decided that it would be best to actively participate in all of the issues related to the library we’re working on. That means reviewing proposals together with the Expensify team and supporting external contributors to ensure quality solutions that won’t introduce any additional regressions. 🙌🏻 

I’m currently going through all of the issues that could require any changes to the library. Feel free to tag me if you need anything, later I’ll be diving into the issues more in-depth, but there’s quite a few of them so it can take a while for me to catch up. Oh and feel free to tag me in any live-markdown related issues in the future! 😄

**Comment by mallenexpensify** [comment]
Moved to Live Markdown project
@kpadmanabhan once https://github.com/Expensify/App/pull/42892 hit production, please hold on proceeding til we have feedback from @BartoszGrajdek .  Thx

**Comment by kpadmanabhan** [comment]
@MelvinBot this is blocked because of the PR that is on hold. 
@jayeshmangwani / @roryabraham : how should we proceed here?

**Comment by jayeshmangwani** [comment]
> how should we proceed here?

@kpadmanabhan We are still waiting for the other contributor to update their PR and get it ready for merging. Until then, we are blocked. 
How should we proceed from here? I am quite confused.
@mallenexpensify  might be the savior here

**Comment by kpadmanabhan** [comment]
> We are still waiting for the other contributor to update their PR and get it ready for merging. Until then, we are blocked.

Exactly. But it has been quite a long wait is my worry.
@mallenexpensify : what do you suggest?

**Comment by mallenexpensify** [comment]
@jayeshmangwani @kpadmanabhan , sorry, I'm confused.  I've checked links but I'm unsure where we're at.  Is it....
- @kpadmanabhan was hired here
- We put it on hold pending https://github.com/Expensify/App/pull/42892
- https://github.com/Expensify/App/pull/42892 never got completed/fixed via the PR, it was fixed elsewhere
- Now we have this open issue, with no action needed.   We need to figure out potential compensation. 

Is that right?

**Comment by kpadmanabhan** [comment]
@mallenexpensify in a sense that's correct.

**Comment by jayeshmangwani** [comment]
> Is that right?

Yes, that's correct! I had opened the discussion a few days ago for this issue here https://expensify.slack.com/archives/C02NK2DQWUX/p1719300668647629

**Comment by mallenexpensify** [comment]
OK.. since @kpadmanabhan was hired and invested a non-immaterial amount of work, compensation should be do. 
My gut says either 50 or 100%.  What do you think @jayeshmangwani and @roryabraham ?

**Comment by jayeshmangwani** [comment]
@mallenexpensify Generally, we should have given 100% as the PR was almost in a mergeable state, but since the raised PR had one regression found during testing, I would say 50% compensation should be given.

**Comment by mallenexpensify** [comment]
Thanks @jayeshmangwani .  Does that sound good n' fair to you @kpadmanabhan ?

**Comment by kpadmanabhan** [comment]
@mallenexpensify though task was for completing the bug fix, I did a lot more than that following up on other bugs and PRs in the related area. However, I am fine with the suggestion from you and @jayeshmangwani . 

Sad that my changes couldn't hit production. Looking forward to that in next proposals.

**Comment by mallenexpensify** [comment]
> I did a lot more than that following up on other bugs and PRs in the related area

@Krishna2323 if you want to provide links and details to the additional work, it's possible that could influence the compensation amount.   We always appreciate contributors monitoring other issues/PRs and chiming in when it's helpful.

**Comment by jayeshmangwani** [comment]
@mallenexpensify tagged the wrong contributor 🙂 ; the message was meant for you, @kpadmanabhan.

**Comment by kpadmanabhan** [comment]
> @kpadmanabhan if you want to provide links and details to the additional work, it's possible that could influence the compensation amount. We always appreciate contributors monitoring other issues/PRs and chiming in when it's helpful.

thanks @mallenexpensify and @jayeshmangwani. 

I am fine with any compensation that Expensify considers fair. 
anyhow noting down my contribution for this issue and the issues that I was keeping track of.

addition of unit tests in my PR for this issue, which i do not see covered in any of the merged PRs from others for this issue.

- https://github.com/Expensify/App/issues/39971#issuecomment-2070625886
- https://github.com/Expensify/App/pull/41265/files

other related issues that i kept track of.

- https://github.com/Expensify/App/pull/42892#issuecomment-2168653907
- https://github.com/Expensify/App/issues/43331#issuecomment-2167821876
- https://github.com/Expensify/App/issues/42216#issuecomment-2140491997
- https://github.com/Expensify/App/issues/40033#issuecomment-2057600733
- https://github.com/Expensify/App/issues/43808#issuecomment-2171064361

not counting the other proposals that i submitted.

**Comment by mallenexpensify** [comment]
> but since the raised PR had one regression found during testing, 

and, from [CONTRIBUTING.md](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md)
> If a PR causes a regression at any point within the regression period (starting when the code is merged and ending 168 hours (that's 7 days) after being deployed to production):

@jayeshmangwani, was the regression found during testing **before** the code was merged?

**Comment by kpadmanabhan** [comment]
> @jayeshmangwani, was the regression found during testing **before** the code was merged?

@mallenexpensify : jayesh found it when everything looked good during final review. 
you may refer to this comment and the surrounding comments in the PR.
https://github.com/Expensify/App/pull/41265#issuecomment-2134547724

correct me if i am wrong @jayeshmangwani

**Comment by jayeshmangwani** [comment]
> was the regression found during testing before the code was merged?

@mallenexpensify Yes, found when testing the PR.

**Comment by mallenexpensify** [comment]
Since the regression was found before being merged, that would be normal testing. We don't have documentation stating there are penalties for that.  Based on that, and @kpadmanabhan 's [breakdown above](https://github.com/Expensify/App/issues/39971#issuecomment-2246110255), I think we shoudl compensate 100% here for both @kpadmanabhan and @jayeshmangwani.  They were both hired/assigned and started work on the PR.

**Comment by mallenexpensify** [comment]
Contributor: @kpadmanabhan  paid $250 via Upwork
Contributor+: @jayeshmangwani paid $250 via Upwork

Thanks!
